bin/oneoff/migrate_old_workshops.rb

```diff
@@ -0,0 +1,166 @@
+  # the db.
+  enrollment.save(validate: false)
+  enrollment.validate!
+
+  return enrollment
+end
+
+Workshop.where(pd_workshop_id: nil).each do |old_workshop|
+  next unless old_workshop.facilitators.count > 0
+
+  new_workshop = Pd::Workshop.new
+
+  # old workshops didn't have organizers, so pick one of the facilitators to be
+  # the new organizer
+  new_workshop.organizer = old_workshop.facilitators.first
+
+  new_workshop.location_name = old_workshop.location
+  new_workshop.created_at = old_workshop.created_at
+
+  # These values don't really matter anymore, but stick em in the notes field
+  # just for posterity
+  new_workshop.notes = "#{old_workshop.name}\n\n#{old_workshop.instructions}"
+
+  # Set Course and Subject based on ProgramType and Phase, which were the old
+  # equivalents.
+  new_workshop.course = PROGRAM_TYPES_TO_COURSES[old_workshop.program_type]
+  if subjects = PHASE_TO_SUBJECT[new_workshop.course]
+    key = old_workshop.phase
+    key = nil unless subjects.key?(key)
+    new_workshop.subject = subjects[key]
+  end
+
+  # capacity doesn't actually matter; set it to a number that will make sure
+  # all the enrollments will fit, and make sure it's not 0
+  new_workshop.capacity = [1, (old_workshop.attendances.count + old_workshop.teachers.count)].max
+
+  # all old workshops were paid but not on the map
+  new_workshop.funded = true
+  new_workshop.on_map = false
+
+  # funded CSF workshops were facilitator-funded
+  if new_workshop.course == Pd::SharedWorkshopConstants::COURSE_CSF
+    new_workshop.funding_type = 'facilitator'
+  end
+
+  new_workshop.save!
+
+  # facilitators => facilitators
+  old_workshop.facilitators.each do |facilitator|
+    new_workshop.facilitators << facilitator
+  end
+
+  # teachers => enrollments
+  old_workshop.teachers.each do |teacher|
+    get_or_create_enrollment(new_workshop, teacher, old_workshop.created_at)
+  end
+
+  # segments => sessions
+  old_workshop.segments.each do |segment|
+    session = Pd::Session.create!(
+      start: segment.start,
+      end: segment.end,
+      created_at: segment.created_at,
+      pd_workshop_id: new_workshop.id
+    )
+
+    # attendances => attendances
+    segment.attendances.each do |old_attendance|
+      next if old_attendance.teacher_id.nil?
+      teacher = old_attendance.teacher
+
+      enrollment = get_or_create_enrollment(new_workshop, teacher, old_attendance.created_at)
+
+      next unless old_attendance.status == "present"
+      next if Pd::Attendance.exists?(pd_session_id: session.id, teacher_id: teacher.id)
+
+      Pd::Attendance.create!(
+        pd_session_id: session.id,
+        teacher_id: teacher.id,
+        pd_enrollment_id: enrollment.id
+      )
+    end
+  end
+
+  new_workshop.save!
+
+  # save the id of the newly-created workshop so we can keep track of which of
+  # the old workshops have and have not been converted.
+  old_workshop.pd_workshop_id = new_workshop.id
+  old_workshop.save!
```

Comment:
@Hamms Any chance we archived these values?  They would come in handy for some additional work I'm doing now.

Reply:
I think they were probably removed by https://github.com/code-dot-org/code-dot-org/pull/26805, unless we have database backups from before then.